File: app/src/main/java/com/dumuzeyn/mp3player/LibraryLoader.java
package com.dumuzeyn.mp3player;

import android.content.Context;
import android.os.Handler;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.RejectedExecutionException;

/** Loads an immutable library snapshot without delaying the Activity's first frame. */
final class LibraryLoader implements AutoCloseable {
    interface Callback {
        void loaded(Snapshot snapshot);
    }

    interface HomeCallback {
        void loaded(HomeContent content);
    }

    static final class Snapshot {
        final ArrayList<Track> tracks;
        final HashSet<String> favorites;
        final ArrayList<Playlist> playlists;
        final HomeContent homeContent;
        final long contentVersion;

        Snapshot(ArrayList<Track> tracks, HashSet<String> favorites,
                ArrayList<Playlist> playlists, long contentVersion) {
            this.tracks = new ArrayList<>(tracks);
            this.favorites = new HashSet<>(favorites);
            this.playlists = new ArrayList<>(playlists);
            this.homeContent = new HomeContentBuilder().build(
                    this.tracks, this.favorites, this.playlists);
            this.contentVersion = contentVersion;
        }
    }


    private final Context context;
    private final Handler mainHandler;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private volatile boolean closed;

    LibraryLoader(Context context, Handler mainHandler) {
        this.context = context;
        this.mainHandler = mainHandler;
    }

    void load(int benchmarkTrackCount, Callback callback) {
        try {
            executor.execute(() -> {
                Snapshot snapshot = readSnapshot(benchmarkTrackCount);
                mainHandler.post(() -> {
                    if (!closed) {
                        callback.loaded(snapshot);
                    }
                });
            });
        } catch (RejectedExecutionException ignored) {
            // Activity is already closing.
        }
    }

    void refreshHome(java.util.Set<String> favorites, java.util.List<Playlist> playlists,
            HomeCallback callback) {
        HashSet<String> favoriteSnapshot = new HashSet<>(favorites);
        ArrayList<Playlist> playlistSnapshot = new ArrayList<>(playlists);
        try {
            executor.execute(() -> {
                LibraryDatabase database = new LibraryDatabase(context.getApplicationContext());
                HomeContent content;
                try {
                    content = new HomeContentBuilder().build(database.loadTracks(),
                            favoriteSnapshot, playlistSnapshot);
                } finally {
                    database.close();
                }
                mainHandler.post(() -> {
                    if (!closed) {
                        callback.loaded(content);
                    }
                });
            });
        } catch (RejectedExecutionException ignored) {
            // Activity is already closing.
        }
    }

    void deriveHome(java.util.List<Track> tracks, java.util.Set<String> favorites,
            java.util.List<Playlist> playlists, HomeCallback callback) {
        ArrayList<Track> trackSnapshot = new ArrayList<>(tracks);
        HashSet<String> favoriteSnapshot = new HashSet<>(favorites);
        ArrayList<Playlist> playlistSnapshot = new ArrayList<>(playlists);
        try {
            executor.execute(() -> {
                HomeContent content = new HomeContentBuilder().build(
                        trackSnapshot, favoriteSnapshot, playlistSnapshot);
                mainHandler.post(() -> {
                    if (!closed) {
                        callback.loaded(content);
                    }
                });
            });
        } catch (RejectedExecutionException ignored) {
            // Activity is already closing.
        }
    }

    @Override
    public void close() {
        closed = true;
        executor.shutdownNow();
    }

    private Snapshot readSnapshot(int benchmarkTrackCount) {
        try {
            Context appContext = context.getApplicationContext();
            if (appContext == null) {
                appContext = context;
            }
            LibraryDatabase.migrateLegacyIfNeeded(appContext);
            BenchmarkLibrarySeeder.seedIfRequested(appContext, benchmarkTrackCount);
            LibraryDatabase database = new LibraryDatabase(appContext);
            try {
                return new Snapshot(database.loadTracks(), database.loadFavorites(),
                        database.loadPlaylists(), LibraryContentVersion.read(appContext));
            } finally {
                database.close();
            }
        } catch (RuntimeException error) {
            VoltuneLog.failure("library_load_failed", error);
            return new Snapshot(new ArrayList<>(), new HashSet<>(), new ArrayList<>(),
                    LibraryContentVersion.read(context));
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/LibraryLoader.kt
package com.dumuzeyn.mp3player

import android.content.Context
import android.os.Handler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/** Loads immutable library snapshots without delaying the Activity's first frame. */
class LibraryLoader(
    private val context: Context,
    private val mainHandler: Handler,
) : AutoCloseable {
    fun interface Callback {
        fun loaded(snapshot: Snapshot)
    }

    fun interface HomeCallback {
        fun loaded(content: HomeContent)
    }

    class Snapshot(
        tracks: ArrayList<Track>,
        favorites: HashSet<String>,
        playlists: ArrayList<Playlist>,
        @JvmField val contentVersion: Long,
    ) {
        @JvmField val tracks = ArrayList(tracks)
        @JvmField val favorites = HashSet(favorites)
        @JvmField val playlists = ArrayList(playlists)
        @JvmField val homeContent: HomeContent =
            HomeContentBuilder().build(this.tracks, this.favorites, this.playlists)
    }

    private val job = SupervisorJob()
    private val scope = CoroutineScope(job + Dispatchers.IO)
    @Volatile private var closed = false

    fun load(benchmarkTrackCount: Int, callback: Callback) {
        scope.launch {
            val snapshot = readSnapshot(benchmarkTrackCount)
            mainHandler.post {
                if (!closed) callback.loaded(snapshot)
            }
        }
    }

    fun refreshHome(
        favorites: Set<String>,
        playlists: List<Playlist>,
        callback: HomeCallback,
    ) {
        val favoriteSnapshot = HashSet(favorites)
        val playlistSnapshot = ArrayList(playlists)
        scope.launch {
            val content = LibraryDatabase(context.applicationContext).use { database ->
                HomeContentBuilder().build(
                    database.loadTracks(),
                    favoriteSnapshot,
                    playlistSnapshot,
                )
            }
            mainHandler.post {
                if (!closed) callback.loaded(content)
            }
        }
    }

    fun deriveHome(
        tracks: List<Track>,
        favorites: Set<String>,
        playlists: List<Playlist>,
        callback: HomeCallback,
    ) {
        val trackSnapshot = ArrayList(tracks)
        val favoriteSnapshot = HashSet(favorites)
        val playlistSnapshot = ArrayList(playlists)
        scope.launch {
            val content = HomeContentBuilder().build(
                trackSnapshot,
                favoriteSnapshot,
                playlistSnapshot,
            )
            mainHandler.post {
                if (!closed) callback.loaded(content)
            }
        }
    }

    override fun close() {
        closed = true
        scope.cancel()
    }

    private fun readSnapshot(benchmarkTrackCount: Int): Snapshot = try {
        val appContext = context.applicationContext ?: context
        LibraryDatabase.migrateLegacyIfNeeded(appContext)
        BenchmarkLibrarySeeder.seedIfRequested(appContext, benchmarkTrackCount)
        LibraryDatabase(appContext).use { database ->
            Snapshot(
                database.loadTracks(),
                database.loadFavorites(),
                database.loadPlaylists(),
                LibraryContentVersion.read(appContext),
            )
        }
    } catch (error: RuntimeException) {
        VoltuneLog.failure("library_load_failed", error)
        Snapshot(
            ArrayList(),
            HashSet(),
            ArrayList(),
            LibraryContentVersion.read(context),
        )
    }
}
